### PR TITLE
refactor healthcheck controller into agentdeploy controller to reduce conflicts

### PR DESCRIPTION
**Pull request by zhiweiyin318**

### CODE REVIEW

File: pkg/addonmanager/controllers/agentdeploy/controller.go [hunk 1]
```diff
@@ -169,6 +172,11 @@ func (c *addonDeployController) sync(ctx context.Context, syncCtx factory.SyncCo
 		return err
 	}
 
+	// to deploy agents after RegistrationApplied is true
+	if !meta.IsStatusConditionTrue(addon.Status.Conditions, addonapiv1alpha1.ManagedClusterAddOnRegistrationApplied) {
```
Comment (qiujian16) [inline pkg/addonmanager/controllers/agentdeploy/controller.go]: we should only check the existence of the condition

File: pkg/addonmanager/controllers/agentdeploy/healthcheck_sync.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+		meta.SetStatusCondition(&addon.Status.Conditions, metav1.Condition{
+			Type:    addonapiv1alpha1.ManagedClusterAddOnConditionAvailable,
+			Status:  metav1.ConditionTrue,
+			Reason:  addonapiv1alpha1.AddonAvailableReasonWorkApply,
+			Message: "Addon manifestWork is applied",
+		})
+		return nil
+	}
+
+	if !meta.IsStatusConditionTrue(addon.Status.Conditions, addonapiv1alpha1.ManagedClusterAddOnManifestApplied) {
```
Comment (qiujian16) [inline pkg/addonmanager/controllers/agentdeploy/healthcheck_sync.go]: I think maybe we should only adding conditions of available when there is manifestapplied condition

**Comment by qiujian16** [comment]
we should also skip AlreadyExists error during create.

**Comment by zhiweiyin318** [comment]
> we should also skip AlreadyExists error during create.

@qiujian16  added

File: pkg/addonmanager/controllers/agentdeploy/controller.go [hunk 3]
```diff
@@ -232,14 +244,54 @@ func (c *addonDeployController) updateAddon(ctx context.Context, new, old *addon
...
+		ctx, new.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+	return err
 }
 
 func (c *addonDeployController) applyWork(ctx context.Context, appliedType string,
 	work *workapiv1.ManifestWork, addon *addonapiv1alpha1.ManagedClusterAddOn) (*workapiv1.ManifestWork, error) {
 
 	work, err := c.workApplier.Apply(ctx, work)
 	if err != nil {
+		if errors.IsAlreadyExists(err) {
```
Comment (qiujian16) [inline pkg/addonmanager/controllers/agentdeploy/controller.go]: will this return "nil, nil"? should we change this in applier instead of here?

Comment (zhiweiyin318) [reply pkg/addonmanager/controllers/agentdeploy/controller.go]: removed the check ,added into applier

**Pull request merged**